modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/logical/LogicalOrToUnionRule.java

```diff
@@ -102,20 +71,94 @@ private LogicalOrToUnionRule(Config config) {
         if (operands == null)
             return;
 
+        if (!idxCollationCheck(call, operands))
+            return;
+
         RelNode input = getInput(call);
 
         RelNode rel0 = createUnionAll(cluster, input, operands.get(0), operands.get(1));
+        RelNode rel1 = createUnionAll(cluster, input, operands.get(1), operands.get(0));
 
-        call.transformTo(rel0, ImmutableMap.of(
-            createUnionAll(cluster, input, operands.get(1), operands.get(0)), rel0
-        ));
+        call.transformTo(rel0, ImmutableMap.of(rel1, rel0));
     }
 
     /** */
-    protected abstract RexNode getCondition(RelOptRuleCall call);
+    private RexNode getCondition(RelOptRuleCall call) {
+        final IgniteLogicalTableScan rel = call.rel(0);
+
+        return rel.condition();
+    }
+
+    /**
+     * Returns common required columns from scan.
+     */
+    private ImmutableBitSet getRequiredColumns(RelOptRuleCall call, int fldCount) {
+        final IgniteLogicalTableScan scan = call.rel(0);
+
+        ImmutableBitSet.Builder builder = ImmutableBitSet.builder();
+
+        Mappings.TargetMapping mapping = Commons.inverseMapping(scan.requiredColumns(), fldCount);
+
+        new RexShuttle() {
+            @Override public RexNode visitLocalRef(RexLocalRef inputRef) {
+                builder.set(mapping.getSourceOpt(inputRef.getIndex()));
+                return inputRef;
+            }
+        }.apply(scan.condition());
+
+        builder.addAll(scan.requiredColumns());
+
+        return builder.build();
+    }
 
     /** */
-    protected abstract RelNode getInput(RelOptRuleCall call);
+    private RelNode getInput(RelOptRuleCall call) {
+        return call.rel(0);
+    }
+
+    /**
+     * @param call Set of appropriate RelNode.
+     * @param operands Operands fron OR expression.
+     *
+     * Compares intersection (currently begining position) of condition and index fields.
+     * This rule need to be triggered only if appropriate indexes will be found otherwise it`s not applicable.
+     */
+    private boolean idxCollationCheck(RelOptRuleCall call, List<RexNode> operands) {
```

Comment:
Seems, we could do better here. Current approach doesn't verify collation properly. It checks only for columns, but misses sorting direction. Please assume follow predicate: `a > 10 AND b < 100`. Assume there are follow indices: `idx_a_desc ON t(a DESC)` and  `idx_b_asc ON t(b ASC)`. This check will pass for cases like above, but in fact it ends up in a full scan rather index range scan.

I would propose to create an index condition for every index and returns as soon as we will find for every operand such an index having non null lower bound

Reply:
done